src/Databases/PostgreSQL/DatabaseMaterializedPostgreSQL.cpp

```diff
@@ -162,8 +249,123 @@ void DatabaseMaterializedPostgreSQL::createTable(ContextPtr local_context, const
+            if (tables_to_replicate.empty())
+                tables_to_replicate = getFormattedTablesList();
+
+            /// tables_to_replicate can be empty if postgres database had no tables when this database was created.
+            SettingChange new_setting("materialized_postgresql_tables_list", tables_to_replicate.empty() ? table_name : (tables_to_replicate + "," + table_name));
+            auto alter_query = createAlterSettingsQuery(new_setting);
+
+            InterpreterAlterQuery(alter_query, current_context).execute();
+
+            auto storage = StorageMaterializedPostgreSQL::create(table, getContext(), remote_database_name, table_name);
+            materialized_tables[table_name] = storage;
+
+            std::lock_guard lock(handler_mutex);
+            replication_handler->addTableToReplication(dynamic_cast<StorageMaterializedPostgreSQL *>(storage.get()), table_name);
+        }
+        catch (...)
+        {
+            /// This is a failed attach table. Remove already created nested table.
+            DatabaseAtomic::dropTable(current_context, table_name, true);
+            throw;
+        }
+    }
+    else
+    {
+        DatabaseAtomic::attachTable(table_name, table, relative_table_path);
+    }
+}
+
+
+StoragePtr DatabaseMaterializedPostgreSQL::detachTable(const String & table_name)
+{
+    /// If there is query context then we need to detach materialized storage.
+    /// If there is no query context then we need to detach internal storage from atomic database.
+    if (CurrentThread::isInitialized() && CurrentThread::get().getQueryContext())
+    {
+        auto & table_to_delete = materialized_tables[table_name];
+        if (!table_to_delete)
+            throw Exception(ErrorCodes::UNKNOWN_TABLE, "Materialized table `{}` does not exist", table_name);
+
+        auto tables_to_replicate = getFormattedTablesList(table_name);
+
+        /// tables_to_replicate can be empty if postgres database had no tables when this database was created.
+        SettingChange new_setting("materialized_postgresql_tables_list", tables_to_replicate);
+        auto alter_query = createAlterSettingsQuery(new_setting);
+
+        {
+            auto current_context = Context::createCopy(getContext()->getGlobalContext());
+            current_context->setInternalQuery(true);
+            InterpreterAlterQuery(alter_query, current_context).execute();
+        }
+
+        auto nested = table_to_delete->as<StorageMaterializedPostgreSQL>()->getNested();
+        if (!nested)
+            throw Exception(ErrorCodes::UNKNOWN_TABLE, "Inner table `{}` does not exist", table_name);
+
+        std::lock_guard lock(handler_mutex);
+        replication_handler->removeTableFromReplication(table_name);
+
+        try
+        {
+            auto current_context = Context::createCopy(getContext()->getGlobalContext());
+            current_context->makeQueryContext();
+            DatabaseAtomic::dropTable(current_context, table_name, true);
```

Comment:
But why we drop it in `detach` method?

Reply:
Because postgres table can be removed from replication via detach only -- for consistency with attach (because we cannot add table to replication via `create table` because we do not know table structure).
So we call detach for materialized table and drop for inner table because this is the behaviour I need.